### PR TITLE
Proof of concept for editor prompt

### PR DESCRIPTION
**Pull request by ChillerDragon**
It works but I did not spend any time on polishing it yet. Would something like this be merged with the design decisions I made? If yes I can finish this up. When finished it would be a big diff and should be merged fast otherwise it is annoying to maintain all the git conflicts.

https://github.com/ddnet/ddnet/assets/20344300/16d5961d-9ee6-47f5-a928-ef2355f9b67f

closes https://github.com/ddnet/ddnet/issues/7933

### CODE REVIEW

**Comment by archimede67** [comment]
I'm sorry I didn't see your mention in #7933. I am not currently working on this feature, however I also had as a project to rework the editor undo/redo to make it more maintainable and integrate this feature alongside it. Now if you already have something working, and you want to polish it, then you can go ahead (make it very pretty :)) and i'll try to integrate my rework later.

**Review by Robyt3 — CHANGES_REQUESTED** [review]
Goes in the right direction. I'd definitely want this to avoid duplicating the button handling code.

I don't think the use of macros is helpful in this PR, I would simply write the code explicitly so it's easier to read.

More comments below in somewhat random order to sketch the solution I would go for.

File: src/game/editor/editor.h [hunk 1]
```diff
@@ -320,11 +323,31 @@ class CEditor : public IEditor
...
 	CLayerSelector *LayerSelector() { return &m_LayerSelector; }
 
+	static void ButtonAddGroup(void *pEditor);
+	static void ButtonRefocus(void *pEditor);
+	static void ButtonProof(void *pEditor);
+	#define REGISTER_BUTTON(index, text, callback, editor) int index;
+	#include <game/editor/buttons.h>
+	#undef REGISTER_BUTTON
+	CEditorButton *m_pButtons;
+	int m_NumButtons;
```
[the remark this answers is not in the document]
Comment (archimede67) [reply src/game/editor/editor.h:333]: I dont think this is quite a great solution, in this example there is only 3 "buttons", but what if we want to add like 50 or 100 different actions? Adding a member manually for each button can become a daunting task. Also, I don't think the name "button" fits, I'd personnally call these "actions"

**Comment by ChillerDragon** [comment]
So whats the status here? Will this be merged if I use robstis approach?

**Comment by archimede67** [comment]
> So whats the status here? Will this be merged if I use robstis approach?

In my opinion that's not the best approach, I don't know if you read the reply to one of Robyt3's comment, but basically this would lead to a huge number of member variables that would need to be manually added, registered, maintained, modified, etc. I am not a big fan of this approach in general.

**Comment by Robyt3** [comment]
General question: Should this feature eventually support multi-step actions? Or is every quick action only associated with exactly one UI button?

I would also go with the name "(quick) action" instead of "button".

> So whats the status here? Will this be merged if I use robstis approach?

Depends on the answer to the question above. If you want to support multi-step actions then this requires some more thinking to properly support them. If you don't, then it should work like this with either approach. I don't like the dynamic memory allocations and usage of integer indices. If you think macros are useful then use them for now, but generate member variables for each individual action using the macros instead of using integer indices and dynamic memory allocation.

**Comment by ChillerDragon** [comment]
> Should this feature eventually support multi-step actions? Or is every quick action only associated with exactly one UI button?

I have not given much thought yet to how many/which features of the editor should be covered by that. What would be a multi step action?

**Comment by Robyt3** [comment]
> I have not given much thought yet to how many/which features of the editor should be covered by that.

You'll need to think about that before we can decide how to implement it.

> What would be a multi step action?

Below some different example for illustration. (They are not necessarily useful to implement.)

Example 1: with fixed steps
1. step: "Add layer"
2. step: "Tiles", "Quads", etc. (this list is static and known at compile time)

Example 2: with user input steps
1. step: "Change grid size"
2. step: "Change grid size: *user inputs size as text*"

Example 3: with dynamic steps
1. step: "Undo"
2. step: *list of all actions that can be undone* (this is dynamic based on the actions performed by the user)

Example 4: contrived example with many steps
1. step: "Start transaction"
2. step: perform several other actions
3. step: "Commit transaction" or "Rollback transaction"

**Comment by ChillerDragon** [comment]
Example 1: I would go with two actions here "Add tile layer", "Add quad layer"
Example 2: Yea that sound nice.
Example 3: Also sounds good.
Example 4: Could be nice but sounds complicated to implement and also to use. Not sure if its worth it.

**Comment by Robyt3** [comment]
I think the current approach can only potentially handle example 1, where all actions are static. To support multi-step actions I would potentially put the current actions and their context information on a stack, so an action could additionally define which actions are available in the next step after it.

**Comment by ChillerDragon** [comment]
The prompt shortcut ctrl+p is conflicting with the proof shortcut. Any chance we can drop the proof shortcut? Or move it to ctrl+shift+p or something like that.

File: src/game/editor/buttons.h [hunk 2]
```diff
@@ -0,0 +1,8 @@
+#ifndef REGISTER_BUTTON
+#define REGISTER_BUTTON(index, text, callback, editor)
+#endif
+
+REGISTER_BUTTON(BTN_ADD_GROUP, "Add group", ButtonAddGroup, this);
+REGISTER_BUTTON(BTN_REFOCUS, "Refocus", ButtonRefocus, this);
+REGISTER_BUTTON(BTN_PROOF, "Proof", ButtonProof, this);
+
```
[the remark this answers is not in the document]
Comment (ChillerDragon) [reply src/game/editor/buttons.h]: I dropped the integer indices and went with member variables. I also added the lambdas. But I will keep the macros for now. Adding one button as of right now requires code changes in 4 places (without macros). That is too much.

File: src/game/editor/buttons.cpp
[the remark this answers is not in the document]
Comment (ChillerDragon) [reply src/game/editor/buttons.cpp:1]: The file is now called `quick_actions.cpp` and implements generic methods like `AddGroup()`. So basically everything that is too bloated for the one liner in the lambda. `editor.cpp` is 9k lines big i think creating a new file with a bunch of small action methods improves the structure.

**Comment by ChillerDragon** [comment]
Current state looks like this and is fully working.

image

File: src/game/editor/editor.cpp [hunk 3]
```diff
@@ -1072,10 +1073,11 @@ void CEditor::DoToolbarLayers(CUIRect ToolBar)
 		// proof button
 		TB_Top.VSplitLeft(40.0f, &Button, &TB_Top);
 		static int s_ProofButton = 0;
-		if(DoButton_Ex(&s_ProofButton, "Proof", MapView()->ProofMode()->IsEnabled(), &Button, 0, "[ctrl+p] Toggles proof borders. These borders represent what a player maximum can see.", IGraphics::CORNER_L) ||
+		CEditorButton Btn = m_pButtons[BTN_PROOF];
+		if(DoButton_Ex(&s_ProofButton, Btn.m_pText, MapView()->ProofMode()->IsEnabled(), &Button, 0, "[ctrl+p] Toggles proof borders. These borders represent what a player maximum can see.", IGraphics::CORNER_L) ||
```
[the remark this answers is not in the document]
Comment (ChillerDragon) [reply src/game/editor/editor.cpp]: Hotkeys do not work while the prompt is open. Also the particular [ctrl-p hotkey is conflicting with the prompt it self](https://github.com/ddnet/ddnet/pull/8358#issuecomment-2211618914). [The tooltip is now shown in the prompt menu](https://github.com/ddnet/ddnet/pull/8358#issuecomment-2211632377).

**Comment by ChillerDragon** [comment]
@Robyt3 implemented all your feedback. Ready for another round of reviews.

File: src/game/editor/prompt.cpp [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+bool CPrompt::OnInput(const IInput::CEvent &Event)
+{
+	if(Input()->KeyIsPressed(KEY_P))
+	{
+		m_Active = true;
+	}
+	return false;
+}
+
+void CPrompt::OnRender(CUIRect View)
```
Comment (Robyt3) [inline src/game/editor/prompt.cpp:48]: Not sure about rendering the prompt over the map view. Should probably be a full `DIALOG_QUICKPROMPT` instead like the file and server-settings dialogs.

**Review by Robyt3 — COMMENTED** [review]
Looks alright, if you only want to handle static actions. It doesn't support the dynamic use cases explained in https://github.com/ddnet/ddnet/pull/8358#issuecomment-2123076025 so far.

**Comment by ChillerDragon** [comment]
> Looks alright, if you only want to handle static actions. It doesn't support the dynamic use cases explained in https://github.com/ddnet/ddnet/pull/8358#issuecomment-2123076025 so far.

Yes I decided against fancy actions. That's okay right?

**Comment by Robyt3** [comment]
> Yes I decided against fancy actions. That's okay right?

Yeah, it's fine by me.

File: src/game/editor/prompt.cpp (hunk 4 above)
Comment (ChillerDragon) [reply src/game/editor/prompt.cpp:48]: Whats wrong with rendering over the map view? Thats seems to be the intended way to use the component system.

**Comment by ChillerDragon** [comment]
Move to https://github.com/ddnet/ddnet/pull/8618

**Pull request closed without merging**